### PR TITLE
fix(5809): test and debug composition divergence with hotkeys.

### LINKED ISSUE #5809: Schema composition: does not diverge with hotkeys
## Design Intention
* Any user changes to the schema composition code block, through the editor text itself (not the schema browser), should diverge (a.k.a. permanently disable) the composition.
* We implement this use the text onChange listener, on the monaco editor.

## Bug:
* this listener does not fire when the hotkey for code commenting-out, is used.

https://user-images.githubusercontent.com/10232835/190710954-f39350c6-7929-48f1-8ec6-f718e4068be9.mov

### PR DESCRIPTION
**Pull request by wiedld**
Closes #5809 

## Bug:
* we found one example where hotkeys were not triggering the composition divergence, while editing the composition block.
  * bug found: commenting out --> was not diverging. 
* goal of this PR is to test the suite of default hotkeys:
  * to make sure we diverge properly.
  * fix it when we don't.

## Approach:
* what we care about per hotkey action, is if there is:
   1. a code change triggered (resulting in a monaco-editor change object).
   2. the change occurs in the composition block .
* since our code is checking the outcomes (e.g. the change object), we can write a test per class of change.
  * hotkey action classes:
    * add lines
    * remove lines
    * move lines, without impacting line counts
    * mutate lines in place.
  * we wrote general test cases for all of these^^, except for the mutate-in-place changes.
* mutate-in-place changes:
  * two flavors, position-specific and non-specific.
  * for positional-specific:
       * we can detect the position of the cursor at the moment of hotkey triggering.
       * e.g. commenting lines
  * for non-specific:
       * we can only look at the monaco-editor change object, which we are already doing.
          * then we use heuristics to determine if the change object is triggered by the LSP-applyEdits.
       * we can have cases where the hotkey action looks exactly like an LSP-applyEdit change.
          * e.g. undo/redo
          *  for now, we cannot handle those.

## Vid:
Commenting out --> now diverges.

https://user-images.githubusercontent.com/10232835/192909443-fad0690c-e511-4b2f-8c66-17e91274911c.mov

## Caveats:
We get diminishing returns pretty quickly, since a lot of our detection of "did this change in the composition block?" relies upon:
* our current approach to identifying the composition block. Meaning, we take line 1 --> line yield.
* trying to diff between LSP applyEdits (w/ composition ownership), versus user changes.

The user can pretty easily make a change which complete destroys our detection using these^^ two factors. We have hardcoded in most of the general cases, but covering all the edge cases is messy. Plus we plan to update the definition of "what is a composition block" and remove the diverge detection, within Q4.

## Checklist
- [x] A PR description, regardless of the triviality of this change, that communicates the value of this PR
- [x] [Well-formatted conventional commit messages](https://www.conventionalcommits.org/en/v1.0.0/) that provide context into the change
- [x] Documentation updated or issue created (provide link to issue/PR)
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [x] ~Feature flagged, if applicable~  N/A

### CODE REVIEW

File: cypress/e2e/shared/fluxQueryBuilder.test.ts [hunk 1]
```diff
@@ -344,7 +345,99 @@ describe('Script Builder', () => {
...
+
+          it('diverges when removing lines', () => {
+            runTest('{cmd+x}')
+          })
+
+          it('diverges when moving lines', () => {
+            runTest('{alt+downArrow}')
+          })
+
+          /// XXX: wiedld (27 Sep 2022) -- we have no way to delineate btwn the LSP applyEdit,
```
[the remark this answers is not in the document]
Comment (wiedld) [reply cypress/e2e/shared/fluxQueryBuilder.test.ts:408]: excellent question.

Here are the order of sub-events, for a schema composition event:
1. executeCommand message is sent to the LSP
2. LSP returns an applyEdit (which we cannot see/detect directly)
3. monaco-editor applies the change
4. we can detect this change using the "change object"

Here are the order of sub-events, for a user triggered hotkey:
1. hotkey pressed
2. editor action is triggered
3. action method is executed
4. monaco-editor applies the change
5. we can then detect the "change object"

The change object has the following information:
* text to replace
* position start/end for replacement

### When we decide to diverge

The LSP applyEdits replace the entire composition block. So we use a heuristic which says that if the "change object" (1) replaces this text, and (2) matches an expected start/end range --> then it's probably a change in response to an LSP requested applyEdit.

Anything which does not match this heuristic --> is probably from the user --> diverge.

This heuristic breaks down with the undo/redo shortcut keys. Because the change object is EXACTLY the same as a previous LSP-triggered change. (Because it's just undoing the last change.)

### Can we fix it? Should we fix it?

The only way we know definitively if the change was requested from the LSP, is if we start tracking our executeCommand requests to the LSP and the LSP applyEdit response. A.k.a. the middleware idea.

But we have decided to not go that route, and to instead: (1) replace any need for "diverge" in Q4, and (2) eventually replace all of the LSP-editor wiring (hopefully Q4). Hence why the unhandled edge case is documented here instead...so we explicitly know what we are not fixing.

File: cypress/e2e/shared/fluxQueryBuilder.test.ts [hunk 2]
```diff
@@ -344,7 +345,99 @@ describe('Script Builder', () => {
...
+            runTest('{shift+alt+downArrow}')
+          })
+
+          it('diverges when removing lines', () => {
+            runTest('{cmd+x}')
+          })
+
+          it('diverges when moving lines', () => {
+            runTest('{alt+downArrow}')
+          })
```
[the remark this answers is not in the document]
Comment (wiedld) [reply cypress/e2e/shared/fluxQueryBuilder.test.ts:406]: Each `it` block does run all of the beforeEach().

These test current run 2 beforeEaches:
* the login
* the clearing/reset of the script

The login can be switched to a before(), but the latter needs to stay a beforeEach()...since we need to reset the script and un-diverge (from the last divergence).

Pushing up the change now, to make the login a single before().

File: src/languageSupport/languages/flux/lsp/connection.ts [hunk 3]
```diff
@@ -134,6 +135,10 @@ class LspConnectionManager {
     })
   }
 
+  /// XXX: wiedld (27 Sep 2022) -- This heuristic is wrong.
+  /// We currently have no way to detect when a change is due to an applyEdit from the LSP.
+  /// versus other changes in the same range, with the same `|> yield(name: "_editor_composition")` included.
+  /// one example is the undo [cmd+z], which "looks" exactly like the previous applyEdit.
```
[the remark this answers is not in the document]
Comment (wiedld) [reply src/languageSupport/languages/flux/lsp/connection.ts:141]: I read these excellent questions, before the lengthy response drafted above ([here](https://github.com/influxdata/ui/pull/5931#discussion_r984022251)).

Will update the in-line comments.

File: cypress/e2e/shared/fluxQueryBuilder.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (wiedld) [reply cypress/e2e/shared/fluxQueryBuilder.test.ts:406]: Experiment confirmed. We do need the login as a beforeEach as well. Otherwise tests will fail. (See monitor-ci responds from single example commit.)

Will leave in both as a beforeEach for now. Total test time has been ~4 minutes (compared to ~8min flows test). Since the divergence feature should hopefully removed by end of Q4, then these tests will also be removed then.

File: src/languageSupport/languages/flux/lsp/connection.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wiedld) [reply src/languageSupport/languages/flux/lsp/connection.ts]: Comment is updated.
Also made the heuristic a more stringent regex. (Which ofc still cannot detect the hotkeys undo/redo.)

File: cypress/e2e/shared/fluxQueryBuilder.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (wiedld) [reply cypress/e2e/shared/fluxQueryBuilder.test.ts:406]: > Wait, don't we want it as a `beforeEach` to isolate each subsequent describe's tests?

Not sure I understand your question. Going to guess at what you meant...and please redirect if I'm missing your point. 😅 

Each test block is an `it` block. So in the case of this:
```
describe('outer scope/context', () => {
    beforeEach(() => {
        console.log('setup outer context')
    })
    describe('inner scope/context 1', () => {
        beforeEach(() => {
            console.log('setup inner context 1')
        })
        it('test A', () => {})
        it('test B', () => {})
    })
    describe('inner scope/context 2', () => {
        it('test C', () => {})
        it('test D', () => {})
    })
})
```

Test A & B will each get the outer and inner scopes run, per each test.
```
'setup outer context'
'setup inner context 1'
'test A'
'setup outer context'
'setup inner context 1'
'test B'
```

The positioning of the beforeEach, whether inside or outside a describe, is NOT about how often the beforeEach runs (since it runs per `it` block). Instead, it's about scoping the context setup.

If we want the "outer scope" beforeEach to run only once per each inner context --> then I believe we have to pull it into each inner scope describe block, and convert it to a before. So it'll be code duplication.

Please feel free to double check this. But it's what I understood from the docs, and then experimenting locally. 🤷🏼
